File: .skills/openclaw-skills/skills/khaney64/discogs-sync/src/discogs_sync/rate_limiter.py
"""Proactive rate limiting via Discogs response headers."""

from __future__ import annotations

import time
import threading


class RateLimiter:
    """Track Discogs rate limit headers and throttle requests proactively."""

    MIN_INTERVAL = 1.1  # seconds between requests (stays under 60/min)
    SLOW_INTERVAL = 2.0  # when remaining <= 5
    PAUSE_DURATION = 10.0  # when remaining <= 2
    LOW_THRESHOLD = 5
    CRITICAL_THRESHOLD = 2

    def __init__(self) -> None:
        self._remaining: int | None = None
        self._last_request_time: float = 0.0
        self._lock = threading.Lock()

    def update_from_headers(self, headers: dict) -> None:
        """Update rate limit state from response headers."""
        remaining = headers.get("X-Discogs-Ratelimit-Remaining")
        if remaining is not None:
            try:
                self._remaining = int(remaining)
            except (ValueError, TypeError):
                pass

    def wait_if_needed(self, verbose: bool = False, description: str = "") -> float:
        """Block until it's safe to make the next request.

        Returns the actual wait time in seconds.
        """
        with self._lock:
            now = time.monotonic()
            elapsed = now - self._last_request_time

            # Determine required interval
            if self._remaining is not None and self._remaining <= self.CRITICAL_THRESHOLD:
                required = self.PAUSE_DURATION
                reason = f"critical (remaining={self._remaining})"
            elif self._remaining is not None and self._remaining <= self.LOW_THRESHOLD:
                required = self.SLOW_INTERVAL
                reason = f"low (remaining={self._remaining})"
            else:
                required = self.MIN_INTERVAL
                reason = "normal"

            wait_time = required - elapsed
            if wait_time > 0:
                if verbose and wait_time > self.MIN_INTERVAL:
                    from .output import print_verbose
                    desc = f" for {description}" if description else ""
                    print_verbose(f"Rate limiter: waiting {wait_time:.1f}s{desc} [{reason}]")
                time.sleep(wait_time)

            self._last_request_time = time.monotonic()
            return max(wait_time, 0.0)

    @property
    def remaining(self) -> int | None:
        return self._remaining


# Global rate limiter instance
_global_limiter = RateLimiter()


def get_rate_limiter() -> RateLimiter:
    return _global_limiter

File: .skills/openclaw-skills/skills/noah-1106/zhihu-fetcher/snippets/rate-limiter.js
/**
 * 频率限制器
 * 简单的令牌桶算法实现
 */

class RateLimiter {
  constructor(minIntervalMs = 2000) {
    this.minInterval = minIntervalMs;
    this.lastRequestTime = 0;
  }

  /**
   * 等待直到可以发送下一个请求
   */
  async wait() {
    const now = Date.now();
    const elapsed = now - this.lastRequestTime;
    
    if (elapsed < this.minInterval) {
      const waitTime = this.minInterval - elapsed;
      console.log(`⏳ 频率限制: 等待 ${waitTime}ms...`);
      await new Promise(resolve => setTimeout(resolve, waitTime));
    }
    
    this.lastRequestTime = Date.now();
  }

  /**
   * 执行带频率限制的函数
   */
  async execute(fn) {
    await this.wait();
    return await fn();
  }
}

/**
 * 全局频率限制器实例
 */
const defaultRateLimiter = new RateLimiter(
  parseInt(process.env.RATE_LIMIT) || 2000
);

module.exports = {
  RateLimiter,
  defaultRateLimiter
};
